### PR TITLE
Clearly document strange header parsing behavior

### LINKED ISSUE #23: Document strange parsing behavior
To document the strange behavior we (@mattiasgrenfeldt and @mihalimara22) have found while testing the current behavior of `net/http` in #5 and #6 we should do the following: For each behavior we have observed that is strange or undesirable we should:

- Create a separate test function that tests this behavior.
- Create one subtest in the function that documents the **current** behavior.
- Create one subtest in the function that documents the **desired** behavior. Use `t.Skip()` so that the tests pass CI.
- Document the behavior using comments and explain why it is strange or undesirable.

(Issue created on @kele 's request.)

### PR DESCRIPTION
**Pull request by mattiasgrenfeldt**
Adresses #23 for the header tests.
All of the tests that show strange or undesirable behavior have now been more clearly documented.

### CODE REVIEW

File: tests/headers/basicauth_test.go [hunk 1]
```diff
@@ -180,3 +163,70 @@ func TestBasicAuth(t *testing.T) {
 		})
 	}
 }
+
+func TestBasicAuthOrdering(t *testing.T) {
+	// It is not clear from the documentation of http.Request.BasicAuth()
```
Comment (mihalimara22) [inline tests/headers/basicauth_test.go:168]: Please rephrase a bit e.g:

The documentation of `http.Request.BasicAuth()` doesn't clearly specify whether only the first Authorization header is used and the other ones are ignored. This could potentially lead to security issues if two HTTP servers that look at different headers are chained together.

The desired behavior would be to decline(?) an HTTP request with more than one Authorization header.

or something along those lines

File: tests/headers/clte_test.go [hunk 2]
```diff
@@ -355,3 +300,229 @@ func TestTransferEncodingChunkSizeLength(t *testing.T) {
 		t.Fatalf("MakeRequest() got err: %v", err)
 	}
 }
+
+func TestMultipleTransferEncodingChunkedFirst(t *testing.T) {
+	// When faced with two Transfer-Encoding headers net/http
```
Comment (mihalimara22) [inline tests/headers/clte_test.go:305]: comma before headers

File: tests/headers/clte_test.go [hunk 3]
```diff
@@ -355,3 +300,229 @@ func TestTransferEncodingChunkSizeLength(t *testing.T) {
...
+	// currently only looks at the first one and ignores the
+	// rest.
+	//
+	// RFC 7230 doesn’t say anything specifically about
```
Comment (mihalimara22) [inline tests/headers/clte_test.go:309]: specific

File: tests/headers/clte_test.go [hunk 4]
```diff
@@ -355,3 +300,229 @@ func TestTransferEncodingChunkSizeLength(t *testing.T) {
...
+	// having multiple TE headers. It only says that in general
```
Comment (mihalimara22) [inline tests/headers/clte_test.go:310]: that, in general, ..

File: tests/headers/clte_test.go [hunk 5]
```diff
@@ -355,3 +300,229 @@ func TestTransferEncodingChunkSizeLength(t *testing.T) {
...
+	//   field value in order, separated by a comma.  The order in
+	//   which header fields with the same field name are received
+	//   is therefore significant to the interpretation of the
+	//   combined field value; a proxy MUST NOT change the order
+	//   of these field values when forwarding a message. [...] “
+	// - RFC 7230 Section 3.2.4
+	//
+	// net/http should probably reject requests with two TE headers
+	// just as it does with requests having two CL headers.
+	// I (grenfeldt@) see no legitimate reason for having two TE
```
Comment (mihalimara22) [inline tests/headers/clte_test.go:326]: I'd replace  "I .." with "There is..."

File: tests/headers/clte_test.go [hunk 6]
```diff
@@ -355,3 +300,229 @@ func TestTransferEncodingChunkSizeLength(t *testing.T) {
...
+		}
+
+		if got, want := extractStatus(resp), statusNotImplemented; got != want {
+			t.Errorf("status code got: %q want: %q", got, want)
+		}
+	})
+}
+
+func TestTransferEncodingListIdentityFirst(t *testing.T) {
+	// In TestTransferEncodingIdentity we see that the
```
Comment (mihalimara22) [inline tests/headers/clte_test.go:460]: I'd avoid pointing the user to the previous test and maybe try to (re)explain the entire issue here.

File: tests/headers/clte_test.go [hunk 7]
```diff
@@ -355,3 +300,229 @@ func TestTransferEncodingChunkSizeLength(t *testing.T) {
...
+	// identity transfer coding should not be supported
+	// at all. But in the current implementation, where
+	// identity is supported, there is a bug. Any transfer
+	// coding that you put after identity in the list of
```
Comment (mihalimara22) [inline tests/headers/clte_test.go:464]: Replace "you" with "the user"

File: tests/headers/host_test.go [hunk 8]
```diff
@@ -111,3 +104,53 @@ func TestHostHeaderMultiple(t *testing.T) {
 		t.Errorf("status code got: %q want: %q", got, want)
 	}
 }
+
+func TestAbsoluteFormURLInvalidSchema(t *testing.T) {
+	// (I (grenfeldt@) don't actually know whether this
```
Comment (mihalimara22) [inline tests/headers/host_test.go:109]: I'd leave out this part and maybe open an issue if you think it's something we should investigate :)

File: tests/headers/referer_test.go [hunk 9]
```diff
@@ -105,3 +93,57 @@ func TestReferer(t *testing.T) {
 		})
 	}
 }
+
+func TestRefererOrdering(t *testing.T) {
+	// It is not clear from the documentation of http.Request.Referer()
+	// that only the first Referer header is used and any other ones
+	// are ignored. This could potentially lead to security issues if two
+	// HTTP servers are chained that look at different headers.
+	//
```
Comment (mihalimara22) [inline tests/headers/referer_test.go:102]: The documentation of http.Request.Referer() doesn't clearly specify whether only the first Referer header is used and the other ones ignored. This could potentially lead to security issues if two HTTP servers that look at different headers are chained together.

File: tests/headers/referer_test.go [hunk 10]
```diff
@@ -105,3 +93,57 @@ func TestReferer(t *testing.T) {
...
+	// The desired behavior would instead be for http.Request.Referer() to
+	// return an error of some sort or an empty string when there are more
```
Comment (mihalimara22) [inline tests/headers/referer_test.go:104]: Can you give an example of some error?

File: tests/headers/useragent_test.go [hunk 11]
```diff
@@ -104,3 +92,57 @@ func TestUserAgent(t *testing.T) {
 		})
 	}
 }
+
+func TestUserAgentOrdering(t *testing.T) {
+	// It is not clear from the documentation of http.Request.UserAgent()
```
Comment (mihalimara22) [inline tests/headers/useragent_test.go:97]: I'd rephrase similarly to what's above as the phrase is not very clear :)

File: tests/headers/basicauth_test.go (hunk 1 above)
Comment (mattiasgrenfeldt) [reply tests/headers/basicauth_test.go]: Thanks, that is more clear! :)
But I don't agree on this phrasing:
`The desired behavior would be to decline(?) an HTTP request with more than one Authorization header.`
We don't want to decline the entire request just because there are more than one Authorization header. We just want the `BasicAuth` function to give an error of some kind (or for the golang team to improve their documentation). Since `BasicAuth` already returns an `ok` bool status I thought we could just set this to `false`.
Or do you think it should be handled differently?

File: tests/headers/referer_test.go (hunk 10 above)
Comment (mattiasgrenfeldt) [reply tests/headers/referer_test.go]: For `Referer()` and `UserAgent()` they don't currently return an error, only a string, so I thought that to signal that something went wrong we might return an empty string. But I think it would be nicer to add an error return value. The kind of error doesn't really matter as long as it is clear what went wrong. So some examples would be:
```go
errors.New("multiple Referer headers")
errors.New("multiple User-Agent headers")
```

Should this amount of detail really be in the comment though? I think that the specific error message would be up to the person implementing this.

File: tests/headers/basicauth_test.go (hunk 1 above)
Comment (empijei) [reply tests/headers/basicauth_test.go]: I think that returning a `Bad Request` status would be acceptable in case of double auth header.
Note that we will likely ban basic auth so the behavior of BasicAuth here is not relevant and [the spec](https://www.w3.org/Protocols/rfc2616/rfc2616-sec14.html#sec14.8) doesn't seem to have an opinion on the matter.

File: tests/headers/referer_test.go (hunk 10 above)
Comment (empijei) [reply tests/headers/referer_test.go]: I think that here we have to do a risk evaluation: getting the referer wrong could have secuirty implications, getting the UA wrong, not too much.

In any case I don't think we should have too strong opinion on the matter as it is normal practice for cooperating clients to not send duplicate request headers, so for our web framework we'll reject other requests, but for the stdlib I guess it is okay to accept odd requests.

File: tests/headers/basicauth_test.go (hunk 1 above)
Comment (mattiasgrenfeldt) [reply tests/headers/basicauth_test.go]: Ok, then I will change the desired behavior to be a `Bad Request`

File: tests/headers/referer_test.go (hunk 10 above)
Comment (mattiasgrenfeldt) [reply tests/headers/referer_test.go]: I'm gonna go with that the desired behavior should be the behavior we want in our framework. I will therefore change the desired behavior for both tests to be to respond with a `400 Bad Request`.

**Comment by mattiasgrenfeldt** [comment]
Rebased on top of master. Ready to merge.

File: tests/headers/basicauth_test.go [hunk 12]
```diff
@@ -180,3 +163,65 @@ func TestBasicAuth(t *testing.T) {
 		})
 	}
 }
+
+func TestBasicAuthOrdering(t *testing.T) {
+	// The documentation of http.Request.BasicAuth() doesn't clearly specify
+	// that only the first Authorization header is used and that the other ones
```
Comment (mihalimara22) [inline tests/headers/basicauth_test.go:169]: Simplify to "the other ones ignored" so it's easier to read :)

File: tests/headers/clte_test.go [hunk 13]
```diff
@@ -355,3 +300,227 @@ func TestTransferEncodingChunkSizeLength(t *testing.T) {
...
 	}
 }
+
+func TestMultipleTransferEncodingChunkedFirst(t *testing.T) {
+	// When faced with two Transfer-Encoding headers, net/http
+	// currently only looks at the first one and ignores the
+	// rest.
+	//
+	// RFC 7230 doesn’t say anything specific about
+	// having multiple TE headers. It only says that, in general
```
Comment (mihalimara22) [inline tests/headers/clte_test.go:310]: missing comma after in general

File: tests/headers/clte_test.go [hunk 14]
```diff
@@ -355,3 +300,227 @@ func TestTransferEncodingChunkSizeLength(t *testing.T) {
...
+		}
+
+		if got, want := extractStatus(resp), statusNotImplemented; got != want {
+			t.Errorf("status code got: %q want: %q", got, want)
+		}
+	})
+}
+
+func TestTransferEncodingListIdentityFirst(t *testing.T) {
+	// There is a bug in the handling of lists of transfer
```
Comment (mihalimara22) [inline tests/headers/clte_test.go:460]: "handling of Transfer-Encoding lists"

File: tests/headers/host_test.go [hunk 15]
```diff
@@ -111,3 +104,51 @@ func TestHostHeaderMultiple(t *testing.T) {
 		t.Errorf("status code got: %q want: %q", got, want)
 	}
 }
+
+func TestAbsoluteFormURLInvalidSchema(t *testing.T) {
+	// When sending a request using the absolute
+	// form as the request target, any schema is currently
+	// accepted.
+	//
+	// The desired behavior would instead be that only
```
Comment (mihalimara22) [inline tests/headers/host_test.go:113]: "to only accept http or https as schemas and to respond with a 400 (Bad Request) when the server receives any other schema"

File: tests/headers/basicauth_test.go (hunk 12 above)
Comment (mattiasgrenfeldt) [reply tests/headers/basicauth_test.go:169]: Do you mean this change?
`and that the other ones are ignored.` -> `and that the other ones ignored.`
That feels less grammatically correct. 
I agree that we can remove the `that` though.

Comment (kele) [inline tests/headers/basicauth_test.go]:
```suggestion
	// that only the first Authorization header is used and the other ones
```
